// @neo4j/graphql/dist/translate/queryAST/ast/filters/aggregation/AggregationFilter.js
"use strict";
/*
 * Copyright (c) "Neo4j"
 * Neo4j Sweden AB [http://neo4j.com]
 *
 * This file is part of Neo4j.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.AggregationFilter = void 0;
const cypher_builder_1 = __importDefault(require("@neo4j/cypher-builder"));
const InterfaceEntityAdapter_1 = require("../../../../../schema-model/entity/model-adapters/InterfaceEntityAdapter");
const context_has_target_1 = require("../../../utils/context-has-target");
const create_node_from_entity_1 = require("../../../utils/create-node-from-entity");
const Filter_1 = require("../Filter");
class AggregationFilter extends Filter_1.Filter {
    constructor(relationship) {
        super();
        this.filters = [];
        this.relationship = relationship;
    }
    addFilters(...filter) {
        this.filters.push(...filter);
    }
    getChildren() {
        return [...this.filters];
    }
    getSubqueries(context) {
        if (!(0, context_has_target_1.hasTarget)(context))
            throw new Error("No parent node found!");
        this.subqueryReturnVariable = new cypher_builder_1.default.Variable();
        const relatedEntity = this.relationship.target;
        const relatedNode = new cypher_builder_1.default.Node();
        let relatedNodeLabels = [];
        let labelsFilter;
        if (relatedEntity instanceof InterfaceEntityAdapter_1.InterfaceEntityAdapter) {
            const labelsForImplementations = relatedEntity.concreteEntities.map((e) => relatedNode.hasLabels(...e.getLabels()));
            labelsFilter = cypher_builder_1.default.or(...labelsForImplementations);
        }
        else {
            relatedNodeLabels = (0, create_node_from_entity_1.getEntityLabels)(relatedEntity, context.neo4jGraphQLContext);
        }
        const relationshipTarget = new cypher_builder_1.default.Relationship();
        const pattern = new cypher_builder_1.default.Pattern(context.target)
            .related(relationshipTarget, {
            direction: this.relationship.getCypherDirection(),
            type: this.relationship.type,
        })
            .to(relatedNode, { labels: relatedNodeLabels });
        const nestedContext = context.push({
            target: relatedNode,
            relationship: relationshipTarget,
        });
        const predicates = cypher_builder_1.default.and(...this.filters.map((f) => f.getPredicate(nestedContext)));
        const returnColumns = [];
        if (predicates) {
            returnColumns.push([predicates, this.subqueryReturnVariable]);
        }
        if (returnColumns.length === 0)
            return []; // Maybe throw?
        const subquery = labelsFilter
            ? new cypher_builder_1.default.Match(pattern).where(labelsFilter).return(...returnColumns)
            : new cypher_builder_1.default.Match(pattern).return(...returnColumns);
        return [subquery];
    }
    getPredicate(_queryASTContext) {
        // should this throw instead?
        if (!this.subqueryReturnVariable)
            return undefined;
        return cypher_builder_1.default.eq(this.subqueryReturnVariable, cypher_builder_1.default.true);
    }
}
exports.AggregationFilter = AggregationFilter;
