// @neo4j/cypher-builder/dist/clauses/Unwind.js
"use strict";
/*
 * Copyright (c) "Neo4j"
 * Neo4j Sweden AB [http://neo4j.com]
 *
 * This file is part of Neo4j.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var Unwind_1;
Object.defineProperty(exports, "__esModule", { value: true });
exports.Unwind = void 0;
const compile_cypher_if_exists_1 = require("../utils/compile-cypher-if-exists");
const Clause_1 = require("./Clause");
const WithCreate_1 = require("./mixins/clauses/WithCreate");
const WithMatch_1 = require("./mixins/clauses/WithMatch");
const WithMerge_1 = require("./mixins/clauses/WithMerge");
const WithReturn_1 = require("./mixins/clauses/WithReturn");
const WithWith_1 = require("./mixins/clauses/WithWith");
const WithDelete_1 = require("./mixins/sub-clauses/WithDelete");
const WithOrder_1 = require("./mixins/sub-clauses/WithOrder");
const WithRemove_1 = require("./mixins/sub-clauses/WithRemove");
const WithSet_1 = require("./mixins/sub-clauses/WithSet");
const Projection_1 = require("./sub-clauses/Projection");
const mixin_1 = require("./utils/mixin");
/**
 * @see [Cypher Documentation](https://neo4j.com/docs/cypher-manual/current/clauses/unwind/)
 * @category Clauses
 */
let Unwind = Unwind_1 = class Unwind extends Clause_1.Clause {
    constructor(...columns) {
        super();
        this.projection = new Projection_1.Projection(columns);
    }
    addColumns(...columns) {
        this.projection.addColumns(columns);
    }
    unwind(clauseOrColumn, ...columns) {
        if (clauseOrColumn instanceof Unwind_1) {
            this.addNextClause(clauseOrColumn);
            return clauseOrColumn;
        }
        return this.addColumnsToUnwindClause(clauseOrColumn, ...columns);
    }
    /** @internal */
    getCypher(env) {
        const projectionStr = this.projection.getCypher(env);
        const deleteCypher = (0, compile_cypher_if_exists_1.compileCypherIfExists)(this.deleteClause, env, { prefix: "\n" });
        const setCypher = (0, compile_cypher_if_exists_1.compileCypherIfExists)(this.setSubClause, env, { prefix: "\n" });
        const removeCypher = (0, compile_cypher_if_exists_1.compileCypherIfExists)(this.removeClause, env, { prefix: "\n" });
        const orderCypher = (0, compile_cypher_if_exists_1.compileCypherIfExists)(this.orderByStatement, env, { prefix: "\n" });
        const nextClause = this.compileNextClause(env);
        return `UNWIND ${projectionStr}${setCypher}${removeCypher}${deleteCypher}${orderCypher}${nextClause}`;
    }
    addColumnsToUnwindClause(...columns) {
        if (!this.nextClause) {
            this.addNextClause(new Unwind_1());
        }
        if (!(this.nextClause instanceof Unwind_1)) {
            throw new Error("Invalid Unwind statement");
        }
        this.nextClause.addColumns(...columns);
        return this.nextClause;
    }
};
exports.Unwind = Unwind;
exports.Unwind = Unwind = Unwind_1 = __decorate([
    (0, mixin_1.mixin)(WithWith_1.WithWith, WithDelete_1.WithDelete, WithMatch_1.WithMatch, WithReturn_1.WithReturn, WithRemove_1.WithRemove, WithSet_1.WithSet, WithCreate_1.WithCreate, WithMerge_1.WithMerge, WithOrder_1.WithOrder)
], Unwind);
