// @neo4j/graphql/dist/schema/constants.js
"use strict";
/*
 * Copyright (c) "Neo4j"
 * Neo4j Sweden AB [http://neo4j.com]
 *
 * This file is part of Neo4j.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.DEPRECATE_DIRECTED_ARGUMENT = exports.DEPRECATE_OPTIONS_ARGUMENT = exports.DEPRECATE_IMPLICIT_EQUAL_FILTERS = void 0;
const constants_1 = require("../constants");
exports.DEPRECATE_IMPLICIT_EQUAL_FILTERS = {
    name: constants_1.DEPRECATED,
    args: {
        reason: "Please use the explicit _EQ version",
    },
};
exports.DEPRECATE_OPTIONS_ARGUMENT = {
    name: constants_1.DEPRECATED,
    args: {
        reason: "Query options argument is deprecated, please use pagination arguments like limit, offset and sort instead.",
    },
};
exports.DEPRECATE_DIRECTED_ARGUMENT = {
    name: constants_1.DEPRECATED,
    args: {
        reason: "The directed argument is deprecated, and the direction of the field will be configured in the GraphQL server",
    },
};
